docs/guides/routing.rst

```diff
@@ -163,40 +163,73 @@ By default, Elgg will set the following requirements for named URL segments:
 	];
 
 
-The ``route`` Plugin Hook
-=========================
+Route middleware
+----------------
 
-The ``route`` plugin hook is triggered before page handlers are called. The URL
-identifier is given as the type of the hook. This hook can be used to add some logic before the
-request is handled elsewhere, or take over page rendering completely.
+Route middleware can be used to prevent access to a certain route, or to perform some business logic before
+the route is rendered. Middleware can be used, e.g. to implement a paywall, or to log analytics,
+or to set open graph metatags.
 
-Generally devs should instead use a page handler unless they need to affect a single page or a wider
-variety of URLs.
+Elgg core implements several middleware handlers, including:
 
-The following code results in ``/blog/all`` requests being completely handled by the plugin hook handler.
-For these requests the ``blog`` page handler is never called.
+ * ``\Elgg\Router\Middleware\Gatekeeper`` - prevent access by non-authenticated users
+ * ``\Elgg\Router\Middleware\AdminGatekeeper`` - prevent access by non-admin users
+ * ``\Elgg\Router\Middleware\AjaxGatekeeper`` - prevent access with non-xhr requests
+ * ``\Elgg\Router\Middleware\CsrfFirewall`` - prevent access without CSRF tokens
 
-.. code-block:: php
+Middleware handlers can be set to any callable that receives an instance of ``\Elgg\Request``:
+The handler should throw an instance of ``HttpException`` to prevent route access.
+The handler can return an instance of ``\Elgg\Http\ResponseBuilder`` to prevent further of the routing sequence.
```

Comment:
Implementation of the routing sequence. Clarify that this can be used to implement custom redirect requests